linkerd/app/gateway/src/lib.rs

```diff
@@ -85,161 +93,246 @@ where
-    svc::stack(http_server.clone())
-        .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Left))
+    // Non-HTTP connections are refused.
+    let legacy_http = inbound
+        .clone()
+        .with_stack(
+            http.clone()
+                .push(svc::NewRouter::layer(RouteHttpLegacy))
+                .push_http_insert_target::<tls::ClientId>(),
+        )
+        .push_http_server()
+        .into_stack()
+        .push_map_target(HttpLegacy::from)
+        .push(svc::UnwrapOr::layer(
+            svc::Fail::<_, RefusedNoTarget>::default(),
+        ))
+        .push(detect::NewDetectService::layer(
+            detect_protocol_timeout,
+            http::DetectHttp::default(),
+        ))
+        .into_inner();
+
+    // When a transported connection is received, use the header's target to
+    // drive routing.
+    inbound
+        .with_stack(
+            http.push_map_target(HttpTarget::from)
+                .push_http_insert_target::<tls::ClientId>(),
+        )
+        .push_http_server()
+        .into_stack()
         .push_switch(
-            |GatewayConnection {
+            |Transported {
                  target,
                  protocol,
                  client,
-             }| match (protocol, target) {
-                (Some(proto), Some(target)) => Ok(svc::Either::A(HttpClientInfo {
-                    target: Some(target),
+             }| match protocol {
+                Some(proto) => Ok(svc::Either::A(HttpTransported {
+                    target,
+                    client,
                     version: match proto {
                         SessionProtocol::Http1 => http::Version::Http1,
                         SessionProtocol::Http2 => http::Version::H2,
                     },
-                    client,
                 })),
-                (None, None) => Ok(svc::Either::B(client)),
-                (None, Some(_)) => Err(Error::from(TcpGatewayUnimplemented(()))),
-                (Some(_), None) => Err(Error::from(RefusedNoTarget(()))),
+                None => Ok::<_, Never>(svc::Either::B(target)),
             },
-            svc::stack(http_server)
-                .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Right))
-                .push_map_target(
-                    |(version, client): (http::Version, ClientInfo)| HttpClientInfo {
-                        target: None,
-                        version,
-                        client,
-                    },
-                )
-                .push(svc::UnwrapOr::layer(
-                    svc::Fail::<_, RefusedNoTarget>::default(),
-                ))
-                .push(detect::NewDetectService::layer(
-                    detect_protocol_timeout,
-                    http::DetectHttp::default(),
-                ))
-                .into_inner(),
+            tcp.into_inner(),
+        )
+        .push_switch(
+            |gw: GatewayConnection| match gw {
```

Comment:
huh, surprised the type can't be inferred from the match? not a big deal if it can't